services/elasticache/pkg/resource/replication_group/custom_update_api.go

```diff
@@ -0,0 +1,388 @@
+// not use this file except in compliance with the License. A copy of the
+// License is located at
+//
+//     http://aws.amazon.com/apache2.0/
+//
+// or in the "license" file accompanying this file. This file is distributed
+// on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+// express or implied. See the License for the specific language governing
+// permissions and limitations under the License.
+
+package replication_group
+
+import (
+	"context"
+	"fmt"
+	ackcompare "github.com/aws/aws-controllers-k8s/pkg/compare"
+	svcapitypes "github.com/aws/aws-controllers-k8s/services/elasticache/apis/v1alpha1"
+	svcsdk "github.com/aws/aws-sdk-go/service/elasticache"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"strconv"
+)
+
+func (rm *resourceManager) UpdateShardConfiguration(
+	ctx context.Context,
+	r *resource,
+	diffReporter *ackcompare.Reporter,
+) (*resource, error) {
+	input, err := rm.newUpdateShardConfigurationRequestPayload(r)
+	if err != nil {
+		return nil, err
+	}
+	resp, respErr := rm.sdkapi.ModifyReplicationGroupShardConfigurationWithContext(ctx, input)
+	if respErr != nil {
+		fmt.Printf("Failed to ModifyReplicationGroupShardConfigurationWithContext, Error: %v\n", respErr)
+		return nil, respErr
+	}
+	return provideUpdatedResource(r, resp.ReplicationGroup)
+}
+
+func (rm *resourceManager) UpdateReplicaCount(
+	ctx context.Context,
+	r *resource,
+	diffReporter *ackcompare.Reporter,
+) (*resource, error) {
+
+	for _, diff := range diffReporter.Differences {
+		if diff.Path == "Spec.ReplicasPerNodeGroup" {
+			desired, err1 := strconv.Atoi(diff.ValueA)
+			latest, err2 := strconv.Atoi(diff.ValueB)
+
+			if err1 != nil || err2 != nil {
+				return nil, fmt.Errorf("UpdateReplicaCount failed: invalid values")
+			}
+
+			if latest < desired { // increase
+				fmt.Printf("Requesting Increase Replica Count. Old value: %v, New Value: %v\n", latest, desired)
+				input, err := rm.newIncreaseReplicaCountRequestPayload(r)
+				if err != nil {
+					fmt.Printf("Error occurred: %v\n", err)
+					return nil, err
+				}
+				resp, respErr := rm.sdkapi.IncreaseReplicaCountWithContext(ctx, input)
+				if respErr != nil {
+					fmt.Printf("Failed to IncreaseReplicaCountWithContext. Error: %v\n", respErr)
+					return nil, respErr
+				}
+				return provideUpdatedResource(r, resp.ReplicationGroup)
+			} else { // decrease
+				input, err := rm.newDecreaseReplicaCountRequestPayload(r)
+				fmt.Printf("Requesting Decrease Replica Count. Old value: %v, New Value: %v\n", latest, desired)
+				if err != nil {
+					fmt.Printf("Error occurred: %v\n", err)
+					return nil, err
+				}
+				resp, respErr := rm.sdkapi.DecreaseReplicaCountWithContext(ctx, input)
+				if respErr != nil {
+					fmt.Printf("Failed to DecreaseReplicaCountWithContext. Error: %v\n", respErr)
+					return nil, respErr
+				}
+				return provideUpdatedResource(r, resp.ReplicationGroup)
+			}
+
+			break
+		}
+	}
```

Comment:
just a stylistic suggestion... you can remove at least one level of indentation in this function by just "searching" for the `DiffItem` and returning an error early if you don't find it. Something like this:

```go
func (rm *resourceManager) UpdateReplicaCount(
	ctx context.Context,
	r *resource,
	diffReporter *ackcompare.Reporter,
) (*resource, error) {

        var di *ackcompare.DiffItem
	for _, diff := range diffReporter.Differences {
		if diff.Path == "Spec.ReplicasPerNodeGroup" {
                        di = diff
			break
		}
	}			
        if di == nil {
                 return nil, fmt.Errorf("expected different ReplicasPerNodeGroup.")
        }

        desired, err1 := strconv.Atoi(diff.ValueA)
        latest, err2 := strconv.Atoi(diff.ValueB)

	if err1 != nil || err2 != nil {
		return nil, fmt.Errorf("UpdateReplicaCount failed: invalid values")
	}

	if latest < desired { // increase
		fmt.Printf("Requesting Increase Replica Count. Old value: %v, New Value: %v\n", latest, desired)
		input, err := rm.newIncreaseReplicaCountRequestPayload(r)
		if err != nil {
			fmt.Printf("Error occurred: %v\n", err)
			return nil, err
		}
		resp, respErr := rm.sdkapi.IncreaseReplicaCountWithContext(ctx, input)
		if respErr != nil {
			fmt.Printf("Failed to IncreaseReplicaCountWithContext. Error: %v\n", respErr)
			return nil, respErr
		}
		return provideUpdatedResource(r, resp.ReplicationGroup)
	}
        // decrease
	input, err := rm.newDecreaseReplicaCountRequestPayload(r)
	fmt.Printf("Requesting Decrease Replica Count. Old value: %v, New Value: %v\n", latest, desired)
	if err != nil {
		fmt.Printf("Error occurred: %v\n", err)
		return nil, err
	}
	resp, respErr := rm.sdkapi.DecreaseReplicaCountWithContext(ctx, input)
	if respErr != nil {
		fmt.Printf("Failed to DecreaseReplicaCountWithContext. Error: %v\n", respErr)
		return nil, respErr
	}
	return provideUpdatedResource(r, resp.ReplicationGroup)
}
```

Reply:
👍 updated code.